api/views.py

```diff
@@ -0,0 +1,87 @@
+import json
+
+from api.models import DeliveryLog
+from api.serializers import DeliveryLogSerializer
+from api.utils import format_destinations, get_default_options
+from django.db import transaction
+from rest_framework.decorators import api_view, permission_classes
+from rest_framework.permissions import IsAuthenticated
+from rest_framework.response import Response
+
+from notification_service.settings import QURIIRI_API_KEY, QURIIRI_API_URL
+from quriiri.send import Sender
+
+sms_sender = Sender(QURIIRI_API_KEY, QURIIRI_API_URL)
+
+
+def validate_data(post_data):
+    pass
+
+
+@api_view(["POST"])
+@permission_classes([IsAuthenticated])
+@transaction.atomic
+def send_message(request):
+    """
+    Payload example
+    {
+      "sender": "Hel.fi",
+      "to": [
+        {
+          "destination": "string",
+          "format": "MOBILE",
+        },
+        {
+          "destination": "string",
+          "format": "MOBILE",
+        },
+        {
+          "destination": "string",
+          "format": "MOBILE",
+        }
+      ],
+      "text": "SMS message"
+    }
+    """
+    data = request.data
+    # TODO: Validate payload
+    validate_data(data)
+
+    log = DeliveryLog.objects.create(user=request.user)
+
+    options = get_default_options(request, id=log.id)
+    destination = format_destinations(data["to"])
+
+    resp = sms_sender.send_sms(data["sender"], destination, data["text"], **options)
+
+    log.report = json.dumps(resp)
+    log.save()
+
+    resp["id"] = str(log.id)
+    return Response(resp)
+
+
+@api_view(["GET"])
+@permission_classes([IsAuthenticated])
+def get_delivery_log(request, id):
+    user = request.user
+    try:
+        log = user.delivery_logs.get(id=id)
+    except DeliveryLog.DoesNotExist:
+        return Response(status=404, data={"error": f" Message ID: {id} does not exist"})
+    data = DeliveryLogSerializer(log).data
+    data["report"] = json.loads(log.report)
+    return Response(data=data)
```

Comment:
Hmm, shouldn't just `return Response(DeliveryLogSerializer(log).data)` work?

Reply:
It wouldn't, we still need to parse the JSON. The Serializer `JSONField` only helps in validating the JSON format while serialize/deserialize the object

Reply:
Ah right, the *model's* field is a `TextField`, if that was a `JSONField` I guess it would work. So next question, have you considered using `JSONField` in the model? :)

Reply:
Now it works! And I don't even need to dumps and loads the json data every time 😆. Thanks!

Reply:
Ok, good 👍  Just to clarify, I didn't necessarily mean that I think you should be using `JSONField` in the model, was just asking have you considered doing so 😄  Now the reports are stored as Postgres native JSONs, which may or may not be what you want, mostly depending on (future?) specs I guess. That said, using those is probably a sane default, and it should be pretty trivial to change them back if ever needed.

Reply:
Well, I couldn't see any reason not to use it at the moment. We don't really care about if it's a JSON or TextField in the DB. The idea of storing the whole Quririi report in a single field in our DB is because we don't want to update the DeliveryLog schema when Quririi updates their log JSON structure, or when we want to support different SMS gateway. Thank Postgres for supporting JSON datatype